starboard/shared/starboard/link_receiver.cc

```diff
@@ -35,77 +41,139 @@ namespace shared {
+      break;
+    case kSbSocketAddressTypeIpv6:
+      socket_fd = socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP);
+      break;
+    default:
+      SB_NOTREACHED();
+      errno = EAFNOSUPPORT;
+      return -1;
+  }
+
+  if (socket_fd < 0) {
+    return -1;
   }
 
-  if (!socket->SetReuseAddress(true)) {
+  // All Starboard sockets are non-blocking, so let's ensure it.
+  if (!posix::SetNonBlocking(socket_fd)) {
+    // Something went wrong, we'll clean up (preserving errno) and return
+    // failure.
+    HANDLE_EINTR(close(socket_fd));
+    return errno;
+  }
+
+#if !defined(MSG_NOSIGNAL) && defined(SO_NOSIGPIPE)
+  // Use SO_NOSIGPIPE to mute SIGPIPE on darwin systems.
+  int optval_set = 1;
+  setsockopt(socket_fd, SOL_SOCKET, SO_NOSIGPIPE,
+             reinterpret_cast<void*>(&optval_set), sizeof(int));
+#endif
+
+  if (socket_fd < 0) {
     SB_LOG(ERROR) << __FUNCTION__ << ": "
-                  << "SbSocketSetReuseAddress failed";
-    return std::unique_ptr<Socket>();
+                  << "Socket create failed, errno: " << errno;
+    return -1;
   }
 
-  return socket;
+  int on = 1;
+  int result = setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  if (result != 0) {
+    SB_LOG(ERROR) << "Failed to set SO_REUSEADDR on socket " << socket_fd
+                  << ", errno = " << errno;
+    return -1;
+  }
+
+  return socket_fd;
 }
 
 // Creates a server socket that is bound to the loopback interface.
-std::unique_ptr<Socket> CreateLocallyBoundSocket(
-    SbSocketAddressType address_type,
-    int port) {
-  std::unique_ptr<Socket> socket = CreateServerSocket(address_type);
-  if (!socket) {
-    return std::unique_ptr<Socket>();
+int CreateLocallyBoundSocket(SbSocketAddressType address_type, int port) {
+  int socket_fd = CreateServerSocket(address_type);
+  if (socket_fd < 0) {
+    return -1;
   }
 
   SbSocketAddress address = {};
   bool success = GetLocalhostAddress(address_type, port, &address);
   if (!success) {
     SB_LOG(ERROR) << "GetLocalhostAddress failed";
-    return std::unique_ptr<Socket>();
+    return -1;
   }
-  SbSocketError result = socket->Bind(&address);
-  if (result != kSbSocketOk) {
-    SB_LOG(ERROR) << __FUNCTION__ << ": "
-                  << "SbSocketBind to " << port << " failed: " << result;
-    return std::unique_ptr<Socket>();
+
+  posix::SockAddr sock_addr;
+  if (!sock_addr.FromSbSocketAddress(&address)) {
+    SB_LOG(ERROR) << __FUNCTION__ << ": Invalid address";
+    return -1;
   }
 
-  return socket;
+  SB_DCHECK(socket_fd >= 0);
+  SbSocketAddress* local_address = &address;
+  // When binding to the IPV6 any address, ensure that the IPV6_V6ONLY flag is
+  // off to allow incoming IPV4 connections on the same socket.
+  // See https://www.ietf.org/rfc/rfc3493.txt for details.
+  if (local_address && (local_address->type == kSbSocketAddressTypeIpv6) &&
+      common::MemoryIsZero(local_address->address, 16)) {
+    int on = 1;
```

Comment:
The comments above state the IPV6_V6ONLY will be turned off. But the way I'm reading this code it is actually turning it on.